### ISSUE TITLE
Building DockerFile fails at Stage 'Install CLIF binary''

### ISSUE THREAD

**Issue by williambrandler** [google/deepvariant#476]
**Have you checked the FAQ? https://github.com/google/deepvariant/blob/r1.2/docs/FAQ.md**: yes

**Describe the issue:**
I am building the deep variant Dockerfile (v1.2) off the Databricks Runtime base image (Ubuntu 18.04).
Run into issues at Stage 'Install CLIF binary', I get the error,

`ModuleNotFoundError: No module named 'apt_pkg'`

I see in the build-prereq.sh script this comment,

`Build clif binary from scratch. Might not be ideal because it installs a
bunch of dependencies, but this works fine when we used this in a Dockerfile
because we don't do build-prereq.sh in the final image.`

Please advise how to get around this when building your own Docker Image

Cheers,

William

**Setup**
  - Operating system: Ununtu 18.04
  - DeepVariant version: 1.2
  - Installation method (Docker, built from source, etc.): Docker
  - Type of data: (sequencing instrument, reference genome, anything special that is unlike the case studies?)

**Steps to reproduce:**
  - Command:
  - Error trace: (if applicable)

**Does the quick start test work on your system?**
Please test with https://github.com/google/deepvariant/blob/r0.10/docs/deepvariant-quick-start.md.
Is there any way to reproduce the issue by using the quick start?

**Any additional context:**

**Comment by danielecook** [comment]
@williambrandler can you provide more detail regarding what you are trying to do?

Have you modified the existing dockerfile to replace the base image with the Databricks runtime image?

The error appears to indicate that the `apt` package manager is not available in the Databricks runtime image.

**Comment by williambrandler** [comment]
hey yes, this is building off the databricks runtime

So replacing,
ARG FROM_IMAGE=ubuntu:20.04
with,
ARG FROM_IMAGE=databricksruntime/minimal:9.x

I also added, 
`sudo -H apt-get -qq -y install python3-apt python3-distutils`

**Comment by danielecook** [comment]
@williambrandler the latest version of DV requires Python 3.8 and Ubuntu 20.04. Are you able to upgrade Ubuntu on the databricks image?

**Comment by williambrandler** [comment]
Not at this time. Are there specific dependencies in Ubuntu 20.04 that are required? Or an older version of DV that has been tested on 18.04?

**Comment by pichuan** [comment]
> Not at this time. Are there specific dependencies in Ubuntu 20.04 that are required? Or an older version of DV that has been tested on 18.04?

@williambrandler You can try v1.1.0, which builds on 18.04: https://github.com/google/deepvariant/tree/r1.1

In v1.2.0, we updated to Python3.8. When I tested this, I was trying to get Python3.8 to work on Ubuntu18.04, but didn't quite get it to work. Not saying it's impossible, but we think it was easiest to update the whole Dockerfile setup to run on Ubuntu20.04 rather than trying to get it to also work on 18.04.


> 
> Would also like some clarification on this statement to help me figure out what is going on,
> 
> `Build clif binary from scratch. Might not be ideal because it installs a bunch of dependencies, but this works fine when we used this in a Dockerfile because we don't do build-prereq.sh in the final image.`

For the build-prereq.sh statement -- what it means is: As part of that script, it runs https://github.com/google/deepvariant/blob/r1.2/tools/build_clif.sh which builds [CLIF](https://github.com/google/clif) from scratch. CLIF is required when you build DeepVariant code. The build_clif.sh script installs a bunch of stuff on your machine. If you directly run ./build-prereq.sh on your machine, you just need to be aware that those things are installed as a side effect.

But, if you're using our Dockerfile and using docker build, we only carry over the built binaries to the next stage. So the extra things installed by CLIF wasn't carried over.

Hopefully that's more clear?

**Comment by williambrandler** [comment]
thanks! 

Ah after some digging I see the newest version of the Databricks Runtime is using Ubuntu 20.04, my bad. 
Let me try again with this version and see how I get on...

**Comment by serge2016** [comment]
Hello! I also need to run DV on Ubuntu 18.04.
But the installation process is too complex. During it python and pip are installed and reinstall several times. I hope you'll provide one linear script for the installation...

**Comment by pichuan** [comment]
Hi @serge2016 , 
to help us understand your use case a bit more, can you tell us why you decided to build from scratch rather than using Docker or Singularity?

Thanks!

**Comment by serge2016** [comment]
Sorry for the delay.
I use DV both from the hos OS and from docker. My host OS is Ubuntu 18.04. And I have many other bioinformatic tools installed, so I can't use the current installation scripts. That's why I'm trying to create another build script, suitable for 18.04 and not conflicting with other python packages.

**Comment by serge2016** [comment]
I hope to make such an installation script soon. How do you think, is it possible for you to change the installation script for mine one if it works? Or there is some kind of building system on your side which is the reason of such complicated and not straight-forward installation?

**Comment by pichuan** [comment]
Hi @serge2016 ,
If you have a setup that works for Ubuntu20.04 as well as Ubuntu18.04, I'd be happy to use it as our default.
You can send a pull request with your changes that works for both Ubuntu20.04 and Ubuntu18.04. Even though we can't directly merge pull requests, I can make corresponding internal changes and mention your contribution in the commit log.

**Comment by pichuan** [comment]
And , @williambrandler , let us know whether you're able to build it with Ubuntu20.04. If you have further questions, feel free to reopen this issue, or open a new one.

**Comment by williambrandler** [comment]
thanks Pi-Chuan, decided to start building from your image with Ubuntu 20.04 to make sure that works before using the Databricks Runtime with Ubuntu 20.04, and got

`#18 0.288 ========== [Tue Aug 10 21:03:43 UTC 2021] Stage 'Install bazel' starting
#18 0.297 ./build-prereq.sh: line 50: bazel: command not found
#18 0.298 ~/bazel /opt/deepvariant
#18 0.298 ./build-prereq.sh: line 56: curl: command not found
------
executor failed running [/bin/sh -c ./build-prereq.sh]: exit code: 127`

I have had no time to test further since then, but plan to get back on this next month

**Comment by pichuan** [comment]
@williambrandler 

The `bazel: command not found` line is likely ok because build-prereq.sh will plan to install bazel after that line.

`curl: command not found` - this might have happened here:

https://github.com/google/deepvariant/blob/r1.2/build-prereq.sh#L75

I'm surprised that at this point there isn't `curl`, because it was installed here:

https://github.com/google/deepvariant/blob/r1.2/build-prereq.sh#L54

You can remove the ` > /dev/null` from that line and see if it was installed currently.

**Comment by williambrandler** [comment]
@pichuan I finally got back to this. I was able to successfully build the container for the ultima genomics fork of deep variant to run on databricks (for CPUs, not tried GPUs yet). 

It did require a few small tweaks to get it working. And I would be happy to run any folks through it if they want to run deep variant on databricks in future...be it on GCP, Azure or AWS

Thanks

**Comment by pichuan** [comment]
Thank you @williambrandler !
Glad to know it works for you. If we (or any other users) need to run DeepVariant on Databricks and encounter any issues, we'll definitely reach out.
If you have some quick tips to share, please feel free to post them here.
